Configurable load test

da-kami opened the pull request:
Note: @Restioson I used branch `test-instrumentation-load-test` to play with the instrumented tests (that branch was on top of your instrumentation code from yesterday).

This PR just contains the load test part; since I removed the `traced_test` library we were initially using one cannot see trace output at the moment. Rebasing this on @Restioson's work wil fix this so I did not bother too much.

I changed the load tests to be more configurable per test and simplified a few things.
The test runner is still somewhat complex, because one can run multiple bobs, multiple protocols and multiple protocol runs in a configurable way. I think this is acceptable, the tests not proof that multiple different scenarios pass under load.

I am not sure how the tests will behave on CI. 
I am not sure if the tests should always be run on CI given that some of them are quite expensive. 
But even if we don't run (all of) them I think we should definitely keep them but potentially ignore some/all. 

Observations: 

- After cleaning up the test setup to be independent of initial connection time the tests pass with high numbers on debug for me. My conclusion so far: Our design should be fine with handling `1000+` takers. 
- The initial connection time changes with the amount of bobs, i.e. the more bobs the more time we need to connect. This is (to some extent) independent of having multiple protocol handlers.

luckysori committed:
Configurable load tests

luckysori commented on xtra-libp2p/tests/load.rs:
```diff
@@ -0,0 +1,392 @@
+    Ok(())
+}
+
+pub fn init_tracing() -> DefaultGuard {
```
🙃 Could move this to `util`.

luckysori commented on xtra-libp2p/tests/load.rs:
```diff
@@ -0,0 +1,392 @@
+}
+
+pub fn init_tracing() -> DefaultGuard {
+    tracing_subscriber::fmt()
+        .with_env_filter("WARN")
+        .with_env_filter("load=debug")
+        .with_env_filter("xtra=debug")
+        .with_env_filter("xtra_libp2p=debug")
+        .with_env_filter("xtras=debug")
+        .with_test_writer()
+        .set_default()
```
🔧 In order to configure `jaeger` and for the logs to also appear on `stdout` I recommend using something like what I did in the `simple-load-test` branch (which is inspired by what we do in `shared_bin::logger`):

https://github.com/itchysats/itchysats/blob/470d6d7f7111384d3e8c1792d750e771a2c87b83/xtra-libp2p/tests/load.rs#L228-L288

The bit that allows us to see the logs is calling `try_init` on the `Registry`. It's different to what we were doing when we were working on these tests together.

da-kami replied:
I purposely created this PR without the Jaeger instrumentation because Cael was still working on it on his fork which makes it a bit hard to depend on without creating conflicts.
But I think these tests are actually independent of the other instrumentation, i.e. these tests have a separate instrumentation setup, so I can bring that commit back into this PR :)
Sorry, did not think that through enough, I will see to add the commit that adds the Jaeger instrumentation.

da-kami replied:
After talking to @Restioson I came to the conclusion that which ever PR comes first, this one or https://github.com/itchysats/itchysats/pull/2404 will require some rebase in the other. 

> But I think these tests are actually independent of the other instrumentation

Counter to my comment above we should actually make use of the macro that is added in #2404 so I don't intend to add the instrumentation in this PR unless #2404 merges first :)

bonomat reviewed (approved):
Good job. Do we intent to merge this and run it as part of the CI? If so, aren't we running into the danger of flaky tests if the CI is under heavy load?

luckysori reviewed (commented):
Just a couple things I've noticed. Haven't reviewed in full yet.

da-kami committed:
Ignore load tests

The load tests are for debug purposes and should be run on CI but can be run locally.
